featurebyte/models/precomputed_lookup_feature_table.py

```diff
@@ -0,0 +1,294 @@
+        lookup_step=entity_lookup_steps[0],
+        feature_store=feature_store,
+    )
+    initial_universe_expr = cast(
+        Select,
+        get_combined_universe(
+            entity_universe_params=[
+                EntityUniverseParams(
+                    graph=lookup_graph_result.graph,
+                    node=lookup_graph_result.lookup_node,
+                    join_steps=None,
+                )
+            ],
+            source_type=feature_store.type,
+        ),
+    )
+    request_table_name = "ENTITY_UNIVERSE"
+    request_table_columns = [
+        SpecialColumnName.POINT_IN_TIME.value,
+        entity_lookup_steps[0].child.serving_name,
+    ]
+    request_expr = expressions.select(
+        expressions.alias_(
+            CURRENT_FEATURE_TIMESTAMP_PLACEHOLDER,
+            alias=SpecialColumnName.POINT_IN_TIME,
+            quoted=True,
+        ),
+        quoted_identifier(entity_lookup_steps[0].child.serving_name),
+    ).from_(initial_universe_expr.subquery())
+    parent_entity_lookup_result = construct_request_table_with_parent_entities(
+        request_table_name=request_table_name,
+        request_table_columns=request_table_columns,
+        join_steps=entity_lookup_steps,
+        feature_store_details=feature_store.get_feature_store_details(),
+    )
+    final_universe_expr = construct_cte_sql(
+        [
+            (request_table_name, request_expr),
+            (
+                parent_entity_lookup_result.new_request_table_name,
+                parent_entity_lookup_result.table_expr,
+            ),
+        ]
+    )
+    final_universe_expr = final_universe_expr.select(
+        *[quoted_identifier(col) for col in parent_entity_lookup_result.new_request_table_columns]
+    ).from_(parent_entity_lookup_result.new_request_table_name)
+    return final_universe_expr
+
+
+def _get_feature_lists_to_relationships_info(
+    feature_lists: List[FeatureListModel],
+) -> Dict[PydanticObjectId, List[EntityRelationshipInfo]]:
+    """
+    Get a mapping from feature list id to the feature list's available relationships info
+
+    Parameters
+    ----------
+    feature_lists: List[FeatureListModel]
+        Feature lists to process
+
+    Returns
+    -------
+    Dict[PydanticObjectId, List[EntityRelationshipInfo]]
+    """
+    feature_lists_relationships_info = {}
+    for feature_list in feature_lists:
+        combined_lookup_steps: List[EntityRelationshipInfo] = []
+        if feature_list.features_entity_lookup_info:
+            for info in feature_list.features_entity_lookup_info:
+                for step in info.join_steps:
+                    if step not in combined_lookup_steps:
+                        combined_lookup_steps.extend(info.join_steps)
+        feature_lists_relationships_info[feature_list.id] = (
+            feature_list.relationships_info or []
+        ) + combined_lookup_steps
+    return feature_lists_relationships_info
```

Comment:
nit: We may consider to move parts of the logic to extract feature list join steps into feature list model later.